File: Archivist/src/test/java/com/zorroa/archivist/web/ExportControllerTests.java
package com.zorroa.archivist.web;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.common.collect.Lists;
import com.zorroa.archivist.repository.ExportDao;
import com.zorroa.archivist.sdk.domain.*;
import com.zorroa.archivist.sdk.processor.ProcessorFactory;
import com.zorroa.archivist.sdk.processor.export.ExportProcessor;
import com.zorroa.archivist.sdk.service.ExportService;
import com.zorroa.archivist.sdk.service.IngestService;
import com.zorroa.archivist.sdk.util.Json;
import com.zorroa.archivist.service.IngestExecutorService;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockHttpSession;
import org.springframework.test.web.servlet.MvcResult;

import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

/**
 * Created by chambers on 11/17/15.
 */
public class ExportControllerTests extends MockMvcTest {

    @Autowired
    ExportService exportService;

    @Autowired
    IngestService ingestService;

    @Autowired
    ExportDao exportDao;

    @Autowired
    IngestExecutorService ingestExecutorService;

    Export export;

    ExportBuilder builder;

    MockHttpSession session;

    @Before
    public void init() {
        session = admin();

        Ingest ingest = ingestService.createIngest(new IngestBuilder(getStaticImagePath()));
        ingestExecutorService.executeIngest(ingest);
        refreshIndex(1000);

        ExportOptions options = new ExportOptions();
        options.getImages().setFormat("jpg");
        options.getImages().setScale(.5);

        AssetSearch search = new AssetSearch().setQuery("beer");

        ProcessorFactory<ExportProcessor> outputFactory = new ProcessorFactory<>();
        outputFactory.setKlass("com.zorroa.archivist.sdk.processor.export.ZipFileExport");

        builder = new ExportBuilder();
        builder.setNote("An export for Bob");
        builder.setOptions(options);
        builder.setSearch(search);
        builder.setOutputs(Lists.newArrayList(outputFactory));

        export = exportService.create(builder);
    }

    @Test
    public void testGetAllByFilter() throws Exception {
        ExportFilter filter = new ExportFilter();

        MvcResult result = mvc.perform(post("/api/v1/exports/_search")
                .session(session)
                .content(Json.serialize(filter))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().isOk())
                .andReturn();

        List<Export> exports = Json.Mapper.readValue(result.getResponse().getContentAsString(),
                new TypeReference<List<Export>>() {});
        assertEquals(1, exports.size());
    }

    @Test
    public void testCreate() throws Exception {
        MvcResult result = mvc.perform(post("/api/v1/exports")
                .session(session)
                .content(Json.serialize(builder))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().isOk())
                .andReturn();

        Export export2 = Json.Mapper.readValue(result.getResponse().getContentAsString(), Export.class);
        assertNotEquals(export.getId(), export2.getId());
        assertEquals("beer", export2.getSearch().getQuery());
        assertEquals(export2.getNote(), export.getNote());
    }

    @Test
    public void testGet() throws Exception {

        MvcResult result = mvc.perform(get("/api/v1/exports/" + export.getId())
                .session(session)
                .content(Json.serialize(builder))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().isOk())
                .andReturn();

        Export export2 = Json.Mapper.readValue(result.getResponse().getContentAsString(), Export.class);
        assertEquals(export.getId(), export2.getId());
        assertEquals("beer", export2.getSearch().getQuery());
        assertEquals(export2.getNote(), export.getNote());
    }

    @Test
    public void testRestart() throws Exception {
        /*
         * artificially set the export to finished.
         */
        exportDao.setState(export, ExportState.Finished, ExportState.Queued);

        MvcResult result = mvc.perform(put("/api/v1/exports/" + export.getId() + "/_restart")
                .session(session)
                .content(Json.serialize(builder))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().isOk())
                .andReturn();

        Export export2 = Json.Mapper.readValue(result.getResponse().getContentAsString(), Export.class);
        assertEquals(ExportState.Queued, export2.getState());
    }

    @Test
    public void testGetAllOutputs() throws Exception {

        MvcResult result = mvc.perform(get("/api/v1/exports/" + export.getId() + "/outputs")
                .session(session)
                .content(Json.serialize(builder))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().isOk())
                .andReturn();

        List<ExportOutput> outputs = Json.Mapper.readValue(result.getResponse().getContentAsString(),
                new TypeReference<List<ExportOutput>>() { });
        assertEquals(1, outputs.size());
    }

    @Test
    public void testCancel() throws Exception {
        /*
         * artificially set the export to running.
         */
        exportDao.setState(export, ExportState.Running, ExportState.Queued);

        MvcResult result = mvc.perform(put("/api/v1/exports/" + export.getId() + "/_cancel")
                .session(session)
                .content(Json.serialize(builder))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().isOk())
                .andReturn();

        Export export2 = Json.Mapper.readValue(result.getResponse().getContentAsString(), Export.class);
        assertEquals(ExportState.Cancelled, export2.getState());
    }
}

File: archivist/src/test/kotlin/com/zorroa/archivist/web/ExportControllerTests.kt
package com.zorroa.archivist.web

import com.zorroa.archivist.domain.ExportFile
import com.zorroa.archivist.domain.ExportFileSpec
import com.zorroa.archivist.domain.ExportSpec
import com.zorroa.archivist.domain.FileStorageSpec
import com.zorroa.archivist.search.AssetSearch
import com.zorroa.archivist.service.ExportService
import com.zorroa.archivist.service.FileStorageService
import com.zorroa.common.domain.Job
import com.zorroa.common.util.Json
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import java.nio.file.Files
import kotlin.test.assertEquals

class ExportControllerTests : MockMvcTest() {

    @Autowired
    lateinit var exportService: ExportService

    @Autowired
    lateinit var fileStorageService: FileStorageService

    @Test
    @Throws(Exception::class)
    fun testCreate() {
        val session = admin()
        addTestAssets("set04/standard")

        val spec = ExportSpec("foo",
                AssetSearch(),
                mutableListOf(),
                mutableMapOf("foo" to "bar"),
                mutableMapOf("foo" to "bar"),
                compress=true)

        val req = mvc.perform(MockMvcRequestBuilders.post("/api/v1/exports")
                .session(session)
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .content(Json.serialize(spec)))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val export = Json.Mapper.readValue<Job>(req.response.contentAsString, Job::class.java)
        assertEquals(spec.name, export.name)
    }


    @Test
    @Throws(Exception::class)
    fun testCreateExportFile() {
        val session = admin()
        addTestAssets("set04/standard")

        val espec = ExportSpec("foo",
                AssetSearch(),
                mutableListOf(),
                mutableMapOf("foo" to "bar"),
                mutableMapOf("foo" to "bar"),
                compress=true)
        val export = exportService.create(espec)

        val storage = fileStorageService.get(FileStorageSpec(
                "export", "foo", "txt", jobId=export.id))
        Files.write(storage.getServableFile().getLocalFile(), "bing".toByteArray())

        val fspec = ExportFileSpec(storage.id, "foo.txt")
        val req = mvc.perform(MockMvcRequestBuilders.post("/api/v1/exports/${export.id}/_files")
                .session(session)
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .content(Json.serialize(fspec)))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val file = Json.Mapper.readValue(req.response.contentAsString, ExportFile::class.java)
        println(Json.prettyString(file))
        assertEquals("export___${export.id}___foo.txt", file.path)
        assertEquals("text/plain", file.mimeType)
        assertEquals(4, file.size)

    }

    @Test
    @Throws(Exception::class)
    fun testGetExportFile() {
        val session = admin()
        addTestAssets("set04/standard")
        this.getTestImagePath()
        val espec = ExportSpec("foo",
                AssetSearch(),
                mutableListOf(),
                mutableMapOf("foo" to "bar"),
                mutableMapOf("foo" to "bar"),
                compress=true)
        val export = exportService.create(espec)


        val storage = fileStorageService.get(FileStorageSpec(
                "export", "foo", "txt", jobId=export.id))
        Files.write(storage.getServableFile().getLocalFile(), "bing".toByteArray())

        val req = mvc.perform(MockMvcRequestBuilders.post("/api/v1/exports/${export.id}/_files")
                .session(session)
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .content(Json.serialize(mapOf("storageId" to storage.id, "filename" to "exported.txt"))))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val file = Json.Mapper.readValue(req.response.contentAsString, ExportFile::class.java)

        val req2 = mvc.perform(MockMvcRequestBuilders.get("/api/v1/exports/${export.id}/_files/${file.id}/_stream")
                .session(session)
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.header().string("Content-Disposition",
                        "attachment; filename=\"exported.txt\""))
                .andExpect(MockMvcResultMatchers.header().longValue("Content-Length",
                        storage.getServableFile().getStat().size))
                .andExpect(MockMvcResultMatchers.header().string("Content-Type", "text/plain"))
                .andReturn()

        val content = req2.response.contentAsString
        assertEquals("bing", content)
    }
}
